rust/src/lib.rs

```diff
@@ -70,39 +83,83 @@ impl fmt::Display for CloudFormationParams {
             "bucket: {}\nkey: {}",
             self.bucket_name,
             match &self.key_arn {
-                None => "None".to_string(),
-                Some(k) => k.to_string(),
+                None => "None",
+                Some(k) => k,
             }
         )
     }
 }
 
+impl Meta {
+    pub fn new_json(algorithm: &str, nonce: &[u8]) -> serde_json::Result<String> {
+        let meta = Meta {
+            alg: algorithm.to_owned(),
+            nonce: base64::engine::general_purpose::STANDARD.encode(nonce),
+        };
+
+        serde_json::to_string(&meta)
+    }
+}
+
+impl S3DataKeys {
+    pub fn new(name: &str) -> S3DataKeys {
+        S3DataKeys {
+            key: format!("{name}.key"),
+            cipher: format!("{name}.aesgcm.encrypted"),
+            meta: format!("{name}.meta"),
+        }
+    }
+
+    /// Return key strings as an array for easy iteration.
+    pub fn as_array(&self) -> [&str; 3] {
+        [&self.key, &self.cipher, &self.meta]
+    }
+
+    /// Convert keys to S3 object identifiers.
+    pub fn to_object_identifiers(&self) -> Vec<ObjectIdentifier> {
+        self.as_array()
+            .iter()
+            .map(|key| {
+                ObjectIdentifier::builder()
+                    .set_key(Some(key.to_string()))
+                    .build()
+                    .unwrap_or_else(|_| panic!("Failed to create ObjectIdentifier for '{key}'"))
+            })
+            .collect()
+    }
+}
+
+impl fmt::Display for Vault {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "region: {}\n{}", self.region, self.cloudformation_params)
+    }
+}
+
 impl Vault {
     pub async fn new(
         vault_stack: Option<&str>,
         region_opt: Option<&str>,
     ) -> Result<Vault, VaultError> {
         let config = aws_config::from_env()
-            .region(get_region_provider(region_opt))
+            .region(Self::get_region_provider(region_opt))
             .load()
             .await;
 
-        // Check env variables directly in case library is not used through the CLI.
-        // These are also handled in the CLI so they are documented in the CLI help.
-        let vault_stack_from_env = get_env_variable("VAULT_STACK");
-        let vault_bucket_from_env = get_env_variable("VAULT_BUCKET");
-        let vault_key_from_env = get_env_variable("VAULT_KEY");
+        // Check env variables directly in case the library is not used through the CLI.
+        // These are also handled in the CLI, so they are documented in the CLI help.
+        let vault_stack_from_env = Self::get_env_variable("VAULT_STACK");
+        let vault_bucket_from_env = Self::get_env_variable("VAULT_BUCKET");
+        let vault_key_from_env = Self::get_env_variable("VAULT_KEY");
```

Comment:
Why encapsulate helper functions inside the `impl Vault`. This makes no sense as those are unrelated to `Vault`. Also involves adding this `Self::`to everything, which just doesn't do anything.

Reply:
These are functions that are only used inside Vault. Rust calls these _associated_ functions, other languages like C++, C#, and Python etc calls them  _static_ functions. The point is to put them inside since they are related to the Vault in that they are only used by it. 

https://doc.rust-lang.org/reference/items/associated-items.html#associated-functions-and-methods

Reply:
`Self::` is needed to reference these functions since they do not operate on an instance of the type, similarly to `new()`.